README.rst

```diff
@@ -0,0 +1,325 @@
+    protected void configure() {
+      setName("PurchaseWorkflow");
+      setDescription("Workflow to parse the purchase events and count the revenue per user and per product");
+
+      addMapReduce("PurchaseEventParser");
+
+      condition(new PurchaseEventVerifier())
+        .fork()
+          .addMapReduce("PurchaseCounterByUser")
+        .also()
+          .addMapReduce("PurchaseCounterByProduct")
+        .join()
+      .otherwise()
+        .addAction(new NotifyByEmail())
+      .end();
+    }
+  }
+
+In the ``configure()`` method we specify the topology for connecting the programs which will run as a part of
+the Workflow execution. For ``PurchaseWorkflow``, after setting the name and description for the Workflow, we add
+MapReduce program ``PurchaseEventParser``. This program will parse the raw purchase events to create the ``Purchase``
+objects from it. After this, we add ``condition`` in the Workflow, which takes a predicate ``PurchaseEventVerifier``.
+If the predicate evaluates to true, we `fork` the execution of the Workflow into two parallel branches.
+One branch executes the ``PurchaseCounterByUser`` MapReduce program, while other executes the
+``PurchaseCounterByProduct`` MapReduce program. If the predicate evaluates to false, then the programs added
+in the ``otherwise`` section will get executed. We have added one dummy custom action, ``EmailNotifier()`` to the
+``otherwise`` section.
+
+Lets take a closer look at the predicate ``PurchaseEventVerifier``.
+
+.. code:: java
+
+  public class PurchaseEventVerifier implements Predicate<WorkflowContext> {
+
+    @Override
+    public boolean apply(WorkflowContext workflowContext) {
+      if (workflowContext == null) {
+        return false;
+      }
+
+      WorkflowToken token = workflowContext.getToken();
+      if (token == null) {
+        return false;
+      }
+
+      Map<String, Map<String, Long>> hadoopCounters = token.getMapReduceCounters();
+      if (hadoopCounters == null) {
+        return false;
+      }
+
+      Map<String, Long> taskCounter = hadoopCounters.get("org.apache.hadoop.mapreduce.TaskCounter");
+
+      if (taskCounter.containsKey("MAP_INPUT_RECORDS")) {
+        long mapInputRecordNumber = taskCounter.get("MAP_INPUT_RECORDS");
+        long mapOutputRecordNumber = taskCounter.get("MAP_OUTPUT_RECORDS");
+        // Return true if at least 80% of the records were processed by previous map job
+        return (mapOutputRecordNumber >= (mapInputRecordNumber * 80/100));
+      }
+
+      return false;
+    }
+  }
+
+``PurchaseEventVerifier`` needs to be a public class which implements the interface ``Predicate<WorkflowContext>``.
+The ``apply`` method in the predicate takes ``WorkflowContext`` as a parameter. The hadoop counters emitted by
+the previous MapReduce program (in our case ``PurchaseEventParser``) can be retrieved in this method using
+``workflowContext`` object. We query for the number of input records to the mapper and number of records
```

Comment:
We query for the number of input records to the mapper and the number of records